[vscode] Support TestRunRequest.preserveFocus

rschnekenbu opened the pull request:
#### What it does

Add a progress service for tests to support activation of views
Support vscode API to listen for test run request preserveFocus:
- when preserve focus is true, nothing had to be done
- if preserveFocus is false, then the Test Results view shall be activated.

fixes #13759

contributed on behalf of STMicroelectronics

#### How to test

1. Start the theia browser example on master and install the following extension (derived from vscode sample extension test-provider-sample):
- src: [test-provider-sample-0.0.1-src.zip](https://github.com/user-attachments/files/15952608/test-provider-sample-0.0.1-src.zip)
- zip vsix: [test-provider-sample-0.0.1.zip](https://github.com/user-attachments/files/15952610/test-provider-sample-0.0.1.zip)

2. When starting extensions from Test UI, nothing should happen (as _preserveFocus_ is always true in this case)
3. With the PR and the extensions, 2 additional commands are available with the markdown tests: Test Sample: Trigger Tests (preserveFocus: **true/false**). With the one with true, nothing shall happen also. The Test Result view shall activate when using the command with the false preserve Focus. 

#### Follow-ups
The current implementation may be extended later based on preferences and test results status. In VS Code, when tests are started or failing, the Test Result view or the Test Explorer view can be activated based on user preferences (see https://github.com/microsoft/vscode/blob/f9238dd94a8b54dc3adb274927c7871e7c215300/src/vs/workbench/contrib/testing/browser/testingProgressUiService.ts#L49)

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

rschnekenbu commented on packages/plugin-ext/src/common/plugin-api-rpc.ts:
Not really sure about naming here. Maybe $notifyTestRunRequested may be more aligned with the other methods?

tsmaeder commented on packages/test/src/browser/view/test-execution-progress-service.ts:
```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import { inject, injectable } from '@theia/core/shared/inversify';
```
Not sure why we need this class. Why isn't this handled in `TestService`?

tsmaeder replied:
Or alternatively, don't go through the test `TestService` but notify this service directly from `test-main.ts`.

rschnekenbu replied:
I had an issue with the architecture there: the test views depend on the TestService, so I could not have the TestService to depend on them.
But yes, I can notify it from test-main.ts. I will propose a new version with TestExecutionProgressService directly used in TestingMain.

rschnekenbu replied:
The new commit has shifted a bit the way it is architectured, as you suggested. TestingMain is directly invoking the progress service and not channeling events through main test service anymore.

tsmaeder commented on packages/plugin-ext/src/plugin/tests.ts:
```diff
@@ -197,7 +198,7 @@ export class TestControllerImpl implements theia.TestController {
+            includeTests, excludeTests, profile, false /* don't support continuous run yet */, preserveFocus
         );
 
         const run = this.testRunStarted(request, name, false, false);
```
We are not notifying the UI with `$notifyOnTestRunRequest`. Why not?

rschnekenbu replied:
The `$notifyOnTestRunRequest` has been replaced by the `$notifyOnTestRunCreated`. It should now be always notified, regardless where the TestRun is started.

tsmaeder commented on packages/plugin/src/theia.d.ts:
```diff
@@ -16242,13 +16242,22 @@ export module '@theia/plugin' {
+         * Controls how test Test Results view is focused.  If true, the editor
+         * will keep the maintain the user's focus. If false, the editor will
+         * prefer to move focus into the Test Results view, although
+         * this may be configured by users.
```
Adding the preference shouldn't be hard, right?

rschnekenbu replied:
The preference has been added, with a bonus additional option: Open Test Explorer on TestRun start.

VS Code has a fourth preference: openOnFail, but i would prefer to focus on other APIs for the end of this month.

rschnekenbu replied:
Default is of course _neverOpen_, I also do not like switching view without explicitly requiring it by hand.

tsmaeder commented on packages/plugin-ext/src/main/browser/test-main.ts:
```diff
@@ -624,4 +625,8 @@ export class TestingMainImpl implements TestingMain {
         this.withController(controllerId).withRun(runId).ended();
     }
+
+    $notifyOnTestRunRequest(preserveFocus: boolean): void {
```
Can't we use `$notifyTestRunCreated` instead of adding a new notification? A test run starts either at creation or when it's "running" flage changes, right?

rschnekenbu replied:
done on new commit.

tsmaeder reviewed (commented):
First reading of the code.

tsmaeder commented on packages/test/src/browser/test-execution-progress-service.ts:
```diff
@@ -0,0 +1,69 @@
+@injectable()
+export class DefaultTestExecutionProgressService implements TestExecutionProgressService {
+
+    @inject(WidgetManager)
```
This seems unused.

tsmaeder commented on packages/test/src/browser/test-execution-progress-service.ts:
```diff
@@ -0,0 +1,69 @@
+    protected readonly widgetManager: WidgetManager;
+
+    @inject(ApplicationShell)
+    protected readonly shell: ApplicationShell;
```
unused

tsmaeder commented on packages/test/src/browser/test-execution-progress-service.ts:
```diff
@@ -0,0 +1,69 @@
+    @inject(TestViewContribution)
+    protected readonly testView: TestViewContribution;
+
+    @inject(TestService)
```
seems unused.

tsmaeder commented:
@rschnekenbu I have a couple of remarks from testing:

1. The preferences are in the "Extensions/Testing" section, but IMO should be in "Features/Testing".
2. When I close the test explorer view, it is not opened when I run the tests with "preserveFocus: false".
3. Not sure what the distinction is between "openOnStart" and "openExplorerOnStart" :shrug:

rschnekenbu commented:
I moved the preferences to right place. That is now under Features section.
I removed the openWithTextExplorer, as the option (copied from vs code) is not really obvious given the comment on the d.ts file and in the description of the preference. Now, the 2 options are 'do nothing' or 'open Test Results View'. I kept however the enum choice if we want to expand later the choices to other views.
I also removed unused code left from previous implementation.

rschnekenbu commented:
A small detail about the test of this feature: Running the tests from the UI should preserve the focus in any case, and not open any views. Only triggering from the extension, with the 'preserveFocus:false' version should open the view if the preference is set for this.

tsmaeder commented:
@rschnekenbu for me revealing the test explorer does not work in this scenario:

1. Set the preference to `openOnTestStart`.
2. Make sure the file explorer view is visible
3. Invoke `Trigger Test (preserveFocus: false)
4. Observe: the test explorer view is not revealed.

🤷

rschnekenbu commented:
> @rschnekenbu for me revealing the test explorer does not work in this scenario:
> 
> 1. Set the preference to `openOnTestStart`.
> 2. Make sure the file explorer view is visible
> 3. Invoke `Trigger Test (preserveFocus: false)
> 4. Observe: the test explorer view is not revealed.
> 
> 🤷

The Test Results View shall open, not the Test Explorer. Is that not the case?

tsmaeder commented:
> The Test Results View shall open, not the Test Explorer. Is that not the case?

No, the file explorer still is visible.

tsmaeder commented:
Cleared this up: The "Test Result View" is the one that shows up in the lower right corner where the terminals, etc. are.
